Segfault checking target

I've been trying to set every thing up, but it didn't work

When requesting http://localhost:9806/?target=unms.example.com I get a segfault in the logs.

Here is how I launch the container and the error It throws (I've redacted the domain and token):

```
# export UNMS_EXPORTER_LOG_LEVEL=debug
# cat /usr/src/unms-exporter/unms_exporter.yml
token: {"unms.example.com": "bla-bla-bla"}
# docker run --name unms_exporter -e UNMS_EXPORTER_LOG_LEVEL \
--volume "/usr/src/unms-exporter/unms_exporter.yml":"/home/unms/unms_exporter.yml" \
--publish 9806:9806 quay.io/ffddorf/unms-exporter \
/usr/local/bin/unms-exporter -c /home/unms/unms_exporter.yml

time="2021-12-25T17:33:03Z" level=info msg="Server starting..." addr="[::]:9806"
time="2021-12-25T17:33:29Z" level=debug msg="Starting request" method=GET url="/?target=unms.example.com"
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xa1390f]

goroutine 23 [running]:
github.com/ffddorf/unms-exporter/exporter.(*Exporter).collectImpl(0xc000339a00, 0xc000128240, 0x0, 0x0)
	/src/exporter/exporter.go:185 +0x76f
github.com/ffddorf/unms-exporter/exporter.(*Exporter).Collect(0xc000339a00, 0xc000128240)
	/src/exporter/exporter.go:115 +0x6c
github.com/prometheus/client_golang/prometheus.(*Registry).Gather.func1()
	/go/pkg/mod/github.com/prometheus/client_golang@v1.9.0/prometheus/registry.go:446 +0x1a2
created by github.com/prometheus/client_golang/prometheus.(*Registry).Gather
	/go/pkg/mod/github.com/prometheus/client_golang@v1.9.0/prometheus/registry.go:538 +0xe8e
```

I have no idea why it fails or how to debug it. 
If I query the unms API form the same machine I get the json data without problem.
`curl --include --header "Accept: application/json" --header "x-auth-token: bla-bla-bla" 'https://unms.example.com/nms/api/v2.1/devices?withInterfaces=true'`.

The uisp server version is 1.3.11 and is running in debian 11.2 with Docker version 20.10.12, build e91ed57.

Thanks in advance.
Alejandroscf

Now I get the error in another line:
```
time="2022-01-04T02:12:31Z" level=info msg="Server starting..." addr="[::]:9806"
time="2022-01-04T02:12:33Z" level=debug msg="Starting request" method=GET url="/?target=unms.example.com"
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0xa14891]

goroutine 27 [running]:
github.com/ffddorf/unms-exporter/exporter.(*Exporter).collectImpl(0xc0002dd9c0, 0xc0004002a0, 0x0, 0x0)
	/src/exporter/exporter.go:164 +0x16f1
github.com/ffddorf/unms-exporter/exporter.(*Exporter).Collect(0xc0002dd9c0, 0xc0004002a0)
	/src/exporter/exporter.go:115 +0x6c
github.com/prometheus/client_golang/prometheus.(*Registry).Gather.func1()
	/go/pkg/mod/github.com/prometheus/client_golang@v1.9.0/prometheus/registry.go:446 +0x1a2
created by github.com/prometheus/client_golang/prometheus.(*Registry).Gather
	/go/pkg/mod/github.com/prometheus/client_golang@v1.9.0/prometheus/registry.go:538 +0xe8e
```
But I managed to fix it in my [commit](https://github.com/alejandroscf/unms-exporter/commit/89c72ee1363e54f8f2f443bcb8674ae1ea53aecc) I don't know if I can append it to the PR

--- a/exporter/exporter.go
+++ b/exporter/exporter.go
@@ -161,7 +161,9 @@ func (e *Exporter) collectImpl(out chan<- prom.Metric) error {
 		out <- prom.MustNewConstMetric(e.metrics["device_maintenance"], prom.GaugeValue, boolToGauge(*device.Meta.Maintenance), deviceLabels...)
 		out <- prom.MustNewConstMetric(e.metrics["device_uptime"], prom.GaugeValue, device.Overview.Uptime, deviceLabels...)
 		out <- prom.MustNewConstMetric(e.metrics["device_last_seen"], prom.CounterValue, float64(time.Time(device.Overview.LastSeen).Unix()), deviceLabels...)
-		out <- prom.MustNewConstMetric(e.metrics["device_last_backup"], prom.GaugeValue, float64(time.Time(*device.LatestBackup.Timestamp).Unix()), deviceLabels...)
+        if device.LatestBackup != nil {
+            out <- prom.MustNewConstMetric(e.metrics["device_last_backup"], prom.GaugeValue, float64(time.Time(*device.LatestBackup.Timestamp).Unix()), deviceLabels...)
+        }
 
 		var wanIF *models.DeviceInterfaceSchema
 		for _, intf := range device.Interfaces {

--- a/exporter/exporter.go
+++ b/exporter/exporter.go
@@ -135,6 +135,10 @@ func boolToGauge(in bool) float64 {
 	return 0
 }
 
+func timeToGauge(ts strfmt.DateTime) float64 {
+	return float64(time.Time(ts).Unix())
+}
+
 var (
 	defaultWithInterfaces = true
 )
@@ -170,10 +174,10 @@ func (e *Exporter) collectImpl(out chan<- prom.Metric) error {
 			out <- prom.MustNewConstMetric(e.metrics["device_cpu"], prom.GaugeValue, device.Overview.CPU, deviceLabels...)
 			out <- prom.MustNewConstMetric(e.metrics["device_ram"], prom.GaugeValue, device.Overview.RAM, deviceLabels...)
 			out <- prom.MustNewConstMetric(e.metrics["device_uptime"], prom.GaugeValue, device.Overview.Uptime, deviceLabels...)
-			out <- prom.MustNewConstMetric(e.metrics["device_last_seen"], prom.CounterValue, float64(time.Time(device.Overview.LastSeen).Unix()), deviceLabels...)
+			out <- prom.MustNewConstMetric(e.metrics["device_last_seen"], prom.CounterValue, timeToGauge(device.Overview.LastSeen), deviceLabels...)
 		}
-		if device.LatestBackup != nil {
-			out <- prom.MustNewConstMetric(e.metrics["device_last_backup"], prom.GaugeValue, float64(time.Time(*device.LatestBackup.Timestamp).Unix()), deviceLabels...)
+		if device.LatestBackup != nil && device.LatestBackup.Timestamp != nil {
+			out <- prom.MustNewConstMetric(e.metrics["device_last_backup"], prom.GaugeValue, timeToGauge(*device.LatestBackup.Timestamp), deviceLabels...)
 		}
 
 		var wanIF *models.DeviceInterfaceSchema

--- a/exporter/exporter.go
+++ b/exporter/exporter.go
@@ -179,20 +179,24 @@ func (e *Exporter) collectImpl(out chan<- prom.Metric) error {
 			)
 
 			out <- prom.MustNewConstMetric(e.metrics["interface_enabled"], prom.GaugeValue, boolToGauge(intf.Enabled), intfLabels...)
-			out <- prom.MustNewConstMetric(e.metrics["interface_plugged"], prom.GaugeValue, boolToGauge(intf.Status.Plugged), intfLabels...)
-			out <- prom.MustNewConstMetric(e.metrics["interface_up"], prom.GaugeValue, boolToGauge(intf.Status.Status == "active"), intfLabels...)
+			if intf.Status != nil {
+				out <- prom.MustNewConstMetric(e.metrics["interface_plugged"], prom.GaugeValue, boolToGauge(intf.Status.Plugged), intfLabels...)
+				out <- prom.MustNewConstMetric(e.metrics["interface_up"], prom.GaugeValue, boolToGauge(intf.Status.Status == "active"), intfLabels...)
+			}
 
-			out <- prom.MustNewConstMetric(e.metrics["interface_dropped"], prom.CounterValue, intf.Statistics.Dropped, intfLabels...)
-			out <- prom.MustNewConstMetric(e.metrics["interface_errors"], prom.CounterValue, intf.Statistics.Errors, intfLabels...)
-			out <- prom.MustNewConstMetric(e.metrics["interface_rx_bytes"], prom.CounterValue, intf.Statistics.Rxbytes, intfLabels...)
-			out <- prom.MustNewConstMetric(e.metrics["interface_tx_bytes"], prom.CounterValue, intf.Statistics.Txbytes, intfLabels...)
-			out <- prom.MustNewConstMetric(e.metrics["interface_rx_rate"], prom.GaugeValue, intf.Statistics.Rxrate, intfLabels...)
-			out <- prom.MustNewConstMetric(e.metrics["interface_tx_rate"], prom.GaugeValue, intf.Statistics.Txrate, intfLabels...)
-			out <- prom.MustNewConstMetric(e.metrics["interface_poe_power"], prom.GaugeValue, intf.Statistics.PoePower, intfLabels...)
+			if intf.Statistics != nil {
+				out <- prom.MustNewConstMetric(e.metrics["interface_dropped"], prom.CounterValue, intf.Statistics.Dropped, intfLabels...)
+				out <- prom.MustNewConstMetric(e.metrics["interface_errors"], prom.CounterValue, intf.Statistics.Errors, intfLabels...)
+				out <- prom.MustNewConstMetric(e.metrics["interface_rx_bytes"], prom.CounterValue, intf.Statistics.Rxbytes, intfLabels...)
+				out <- prom.MustNewConstMetric(e.metrics["interface_tx_bytes"], prom.CounterValue, intf.Statistics.Txbytes, intfLabels...)
+				out <- prom.MustNewConstMetric(e.metrics["interface_rx_rate"], prom.GaugeValue, intf.Statistics.Rxrate, intfLabels...)
+				out <- prom.MustNewConstMetric(e.metrics["interface_tx_rate"], prom.GaugeValue, intf.Statistics.Txrate, intfLabels...)
+				out <- prom.MustNewConstMetric(e.metrics["interface_poe_power"], prom.GaugeValue, intf.Statistics.PoePower, intfLabels...)
+			}
 		}
 
 		// WAN metrics
-		if wanIF != nil {
+		if wanIF != nil && wanIF.Statistics != nil {
 			out <- prom.MustNewConstMetric(e.metrics["wan_rx_bytes"], prom.CounterValue, wanIF.Statistics.Rxbytes, deviceLabels...)
 			out <- prom.MustNewConstMetric(e.metrics["wan_tx_bytes"], prom.CounterValue, wanIF.Statistics.Txbytes, deviceLabels...)
 			out <- prom.MustNewConstMetric(e.metrics["wan_rx_rate"], prom.GaugeValue, wanIF.Statistics.Rxrate, deviceLabels...)

--- a/exporter/exporter.go
+++ b/exporter/exporter.go
@@ -155,12 +155,16 @@ func (e *Exporter) collectImpl(out chan<- prom.Metric) error {
 			device.Identification.Site.Name, // siteName
 		}
 
-		out <- prom.MustNewConstMetric(e.metrics["device_cpu"], prom.GaugeValue, device.Overview.CPU, deviceLabels...)
-		out <- prom.MustNewConstMetric(e.metrics["device_ram"], prom.GaugeValue, device.Overview.RAM, deviceLabels...)
 		out <- prom.MustNewConstMetric(e.metrics["device_enabled"], prom.GaugeValue, boolToGauge(*device.Enabled), deviceLabels...)
-		out <- prom.MustNewConstMetric(e.metrics["device_maintenance"], prom.GaugeValue, boolToGauge(*device.Meta.Maintenance), deviceLabels...)
-		out <- prom.MustNewConstMetric(e.metrics["device_uptime"], prom.GaugeValue, device.Overview.Uptime, deviceLabels...)
-		out <- prom.MustNewConstMetric(e.metrics["device_last_seen"], prom.CounterValue, float64(time.Time(device.Overview.LastSeen).Unix()), deviceLabels...)
+		if device.Meta != nil {
+			out <- prom.MustNewConstMetric(e.metrics["device_maintenance"], prom.GaugeValue, boolToGauge(*device.Meta.Maintenance), deviceLabels...)
+		}
+		if device.Overview != nil {
+			out <- prom.MustNewConstMetric(e.metrics["device_cpu"], prom.GaugeValue, device.Overview.CPU, deviceLabels...)
+			out <- prom.MustNewConstMetric(e.metrics["device_ram"], prom.GaugeValue, device.Overview.RAM, deviceLabels...)
+			out <- prom.MustNewConstMetric(e.metrics["device_uptime"], prom.GaugeValue, device.Overview.Uptime, deviceLabels...)
+			out <- prom.MustNewConstMetric(e.metrics["device_last_seen"], prom.CounterValue, float64(time.Time(device.Overview.LastSeen).Unix()), deviceLabels...)
+		}
 		if device.LatestBackup != nil {
 			out <- prom.MustNewConstMetric(e.metrics["device_last_backup"], prom.GaugeValue, float64(time.Time(*device.LatestBackup.Timestamp).Unix()), deviceLabels...)
 		}

--- a/exporter/exporter.go
+++ b/exporter/exporter.go
@@ -121,6 +121,13 @@ func (e *Exporter) Collect(out chan<- prom.Metric) {
 	}
 }
 
+func derefOrFalse(in *bool) bool {
+	if in == nil {
+		return false
+	}
+	return *in
+}
+
 func boolToGauge(in bool) float64 {
 	if in {
 		return 1
@@ -155,9 +162,9 @@ func (e *Exporter) collectImpl(out chan<- prom.Metric) error {
 			device.Identification.Site.Name, // siteName
 		}
 
-		out <- prom.MustNewConstMetric(e.metrics["device_enabled"], prom.GaugeValue, boolToGauge(*device.Enabled), deviceLabels...)
+		out <- prom.MustNewConstMetric(e.metrics["device_enabled"], prom.GaugeValue, boolToGauge(derefOrFalse(device.Enabled)), deviceLabels...)
 		if device.Meta != nil {
-			out <- prom.MustNewConstMetric(e.metrics["device_maintenance"], prom.GaugeValue, boolToGauge(*device.Meta.Maintenance), deviceLabels...)
+			out <- prom.MustNewConstMetric(e.metrics["device_maintenance"], prom.GaugeValue, boolToGauge(derefOrFalse(device.Meta.Maintenance)), deviceLabels...)
 		}
 		if device.Overview != nil {
 			out <- prom.MustNewConstMetric(e.metrics["device_cpu"], prom.GaugeValue, device.Overview.CPU, deviceLabels...)

--- a/exporter/exporter.go
+++ b/exporter/exporter.go
@@ -161,9 +161,9 @@ func (e *Exporter) collectImpl(out chan<- prom.Metric) error {
 		out <- prom.MustNewConstMetric(e.metrics["device_maintenance"], prom.GaugeValue, boolToGauge(*device.Meta.Maintenance), deviceLabels...)
 		out <- prom.MustNewConstMetric(e.metrics["device_uptime"], prom.GaugeValue, device.Overview.Uptime, deviceLabels...)
 		out <- prom.MustNewConstMetric(e.metrics["device_last_seen"], prom.CounterValue, float64(time.Time(device.Overview.LastSeen).Unix()), deviceLabels...)
-        if device.LatestBackup != nil {
-            out <- prom.MustNewConstMetric(e.metrics["device_last_backup"], prom.GaugeValue, float64(time.Time(*device.LatestBackup.Timestamp).Unix()), deviceLabels...)
-        }
+		if device.LatestBackup != nil {
+			out <- prom.MustNewConstMetric(e.metrics["device_last_backup"], prom.GaugeValue, float64(time.Time(*device.LatestBackup.Timestamp).Unix()), deviceLabels...)
+		}
 
 		var wanIF *models.DeviceInterfaceSchema
 		for _, intf := range device.Interfaces {

--- a/exporter/exporter.go
+++ b/exporter/exporter.go
@@ -157,6 +157,9 @@ func (e *Exporter) collectImpl(out chan<- prom.Metric) error {
 	}
 
 	for _, device := range devices.Payload {
+		if device.Identification == nil {
+			continue
+		}
 		deviceLabels := []string{
 			*device.Identification.ID,       // deviceId
 			device.Identification.Name,      // deviceName
@@ -182,6 +185,10 @@ func (e *Exporter) collectImpl(out chan<- prom.Metric) error {
 
 		var wanIF *models.DeviceInterfaceSchema
 		for _, intf := range device.Interfaces {
+			if intf.Identification == nil {
+				continue
+			}
+
 			if intf.Identification.Name == device.Identification.WanInterfaceID {
 				wanIF = intf
 			}

